app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java

```diff
@@ -17,37 +9,38 @@
 import fr.free.nrw.commons.utils.ImageUtils;
 import io.reactivex.Single;
 
+/**
+ * We want to discourage users from uploading images to Commons that were taken from Facebook.
+ * This attempts to detect whether an image was downloaded from Facebook by heuristically
+ * searching for metadata that is specific to images that come from Facebook.
+ */
 @Singleton
 public class ReadFBMD {
 
     @Inject
     public ReadFBMD() {
-
     }
 
-    public Single<Integer> processMetadata(Context context, Uri contentUri) throws IOException {
-        Metadata readMetadata = null;
+    public Single<Integer> processMetadata(String path) {
         try {
-            readMetadata = ImageMetadataReader.readMetadata(context.getContentResolver().openInputStream(contentUri));
-        } catch (ImageProcessingException e) {
-            e.printStackTrace();
-        } catch (IOException e) {
-            e.printStackTrace();
-        }
+            int psBlockOffset;
+            int fbmdOffset;
+
+            try (FileInputStream fs = new FileInputStream(path)) {
+                byte[] bytes = new byte[4096];
+                fs.read(bytes);
+                fs.close();
+                String fileStr = new String(bytes);
+                psBlockOffset = fileStr.indexOf("8BIM");
+                fbmdOffset = fileStr.indexOf("FBMD");
+            }
 
-        IptcDirectory iptcDirectory = readMetadata != null ? readMetadata.getFirstDirectoryOfType(IptcDirectory.class) : null;
-        if (iptcDirectory == null) {
-            return Single.just(ImageUtils.IMAGE_OK);
-        }
-        /**
-         * We parse through all the tags in the IPTC directory  if the tagname equals "Special Instructions".
-         * And the description string starts with FBMD.
-         * Then the source of image is facebook
-         * */
-        for (Tag tag : iptcDirectory.getTags()) {
-            if (tag.getTagName().equals("Special Instructions") && tag.getDescription().substring(0, 4).equals("FBMD")) {
+            if (psBlockOffset > 0 && fbmdOffset > 0
+                    && fbmdOffset > psBlockOffset && fbmdOffset - psBlockOffset < 0x80) {
```

Comment:
While removing a 656 KB library is great, in my humble opinion I would say that this implementation is much less readable, replacing easy-to-read business logic with byte manipulation and unexplained magic codes such as "8BIM" and 0x80.

Reply:
Maybe some comments/explanations could help. In that case would you still think this is less readable @nicolas-raoul ?

Reply:
@dbrant is it possible to add some comments here to make it more readable?